File: Yogscart Alpha/Assets/Scripts/Game Modes/SortingScript.js
#pragma strict
import System.Collections.Generic; //Cause lazy
import System.Linq; //Cause lazy

function CalculatePositions(array : Racer[])
{

	var sortedArray = new Racer[0];
	var finished : int = 0;
	var copy = new Array();

	for(var i : int = 0; i < array.Length; i++)
	{
		if(!array[i].finished)
			copy.Push(array[i]);
		else
			finished += 1;
	}

	sortedArray = copy;

	quickSort(sortedArray,0,sortedArray.Length-1);

	for(i = finished; i < array.Length; i++)
	{
	sortedArray[i-finished].position = i;
}

}

function quickSort(array : Racer[],left : int, right : int)
{
	if(right - left >= 1)
	{

	//Debug.Log("Quick Sorting between " + left + " and " + right);

	var pivot : int = (right + left)/2;
	var leftCheck : int = left;
	var rightCheck : int = right;

	while(leftCheck < rightCheck)
	{
		while(array[leftCheck].TotalDistance > array[pivot].TotalDistance || 
			(array[leftCheck].TotalDistance == array[pivot].TotalDistance && array[leftCheck].NextDistance < array[pivot].NextDistance))
				leftCheck += 1;
				
		while(array[rightCheck].TotalDistance < array[pivot].TotalDistance || 
			(array[rightCheck].TotalDistance == array[pivot].TotalDistance && array[rightCheck].NextDistance > array[pivot].NextDistance))
				rightCheck -= 1;
				
		if(leftCheck < rightCheck)
		{
			//Debug.Log("Swapping " + leftCheck.ToString() + " & " + rightCheck.ToString());
			
			if(leftCheck == pivot)
			{
				pivot = rightCheck;	
				//Debug.Log("pivot has swapped " + pivot.ToString());	
			}
			else if(rightCheck == pivot)
			{
				pivot = leftCheck;
				//Debug.Log("pivot has swapped " + pivot.ToString());	
			}
			
			Swap(array,leftCheck,rightCheck);
			
			if(leftCheck != pivot)
			leftCheck += 1;
			
			if(rightCheck != pivot)
			rightCheck -= 1;

		}


	}

	quickSort(array,left,pivot - 1);
	quickSort(array,pivot + 1,right);

	}

return;
	
}

function Swap(array : Racer[], a : int, b : int)
{

	var holder = array[a];

	array[a] = array[b];
	array[b] = holder;

	return;

}

static function SortRacersPoints(toChangeArray : Racer[])
{

	var array = new List.<Racer>();
	
	for(var j : int = 0; j < toChangeArray.Length; j++)
		array.Add(toChangeArray[j]);

	var sorted : boolean = false;
	var endInt : int = 0;

	while(!sorted)
	{
	
		sorted = true;
					
		for(var i : int = 1; i < array.Count - endInt; i++)
		{
			if(array[i-1].points < array[i].points)
			{
				
				var holder = array[i-1];
				array[i-1] = array[i];
				array[i] = holder;
				sorted = false;
			}
		}
		
		endInt++;
	}
	
	return array;

}


static function CalculatePoints(toChangeArray : List.<DisplayRacer>)
{

	var array = new List.<DisplayRacer>();
	
	for(var j : int = 0; j < toChangeArray.Count; j++)
		array.Add(toChangeArray[j]);

	var sorted : boolean = false;
	var endInt : int = 0;

	while(!sorted)
	{
	
		sorted = true;
					
		for(var i : int = 1; i < array.Count - endInt; i++)
		{
			if(array[i-1].points < array[i].points)
			{
				
				var holder = array[i-1];
				array[i-1] = array[i];
				array[i] = holder;
				sorted = false;
			}
		}
		
		endInt++;
	}
	
	return array;

}

File: Yogscart C#/Assets/Scripts/Game Data/SortingScript.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System;

public class SortingScript : MonoBehaviour
{

    static public void CalculatePositions(List<Racer> array)
    {
        int finished = 0;
        List<Racer> sortedArray = new List<Racer>();

        for (int i = 0; i < array.Count; i++)
        {
            if (!array[i].finished)
                sortedArray.Add(array[i]);
            else
                finished += 1;
        }

        QuickSort(sortedArray, 0, sortedArray.Count - 1);

        for (int i = 0; i < sortedArray.Count; i++)
        {
            sortedArray[i].position = finished + i;
        }

        //Check if two racers have the same position
        List<int> collectedPositions = new List<int>();
        for (int i = 0; i < array.Count; i++)
        {
            if (!collectedPositions.Contains(array[i].position))
            {
                collectedPositions.Add(array[i].position);
            }
            else
            {
                throw new Exception("Same position appears twice!!!");
            }
        }

    }

    static private void QuickSort(List<Racer> array, int left, int right)
    {
        if (right - left >= 1)
        {
            int pivot = (right + left) / 2, leftCheck = left, rightCheck = right;

            while (leftCheck < rightCheck)
            {
                while (array[leftCheck].totalDistance > array[pivot].totalDistance ||
                    (array[leftCheck].totalDistance == array[pivot].totalDistance && array[leftCheck].currentDistance < array[pivot].currentDistance))
                    leftCheck += 1;

                while (array[rightCheck].totalDistance < array[pivot].totalDistance ||
                    (array[rightCheck].totalDistance == array[pivot].totalDistance && array[rightCheck].currentDistance > array[pivot].currentDistance))
                    rightCheck -= 1;

                if (leftCheck < rightCheck)
                {
                    if (leftCheck == pivot)
                    {
                        pivot = rightCheck;
                    }
                    else if (rightCheck == pivot)
                    {
                        pivot = leftCheck;
                    }

                    Swap(array, leftCheck, rightCheck);

                    if (leftCheck != pivot)
                        leftCheck += 1;

                    if (rightCheck != pivot)
                        rightCheck -= 1;

                }


            }

            QuickSort(array, left, pivot - 1);
            QuickSort(array, pivot + 1, right);

        }

        return;
    }

    static public List<DisplayRacer> CalculatePoints(List<DisplayRacer> toSort)
    {
        bool sorted = false;
        int endInt = 0;

        List<DisplayRacer> array = new List<DisplayRacer>(toSort);

        while (!sorted)
        {
            sorted = true;

            for (int i = 1; i < array.Count - endInt; i++)
            {
                if (array[i - 1].points < array[i].points)
                {
                    DisplayRacer holder = array[i - 1];
                    array[i - 1] = array[i];
                    array[i] = holder;
                    sorted = false;
                }
            }

            endInt++;
        }

        return array;

    }

    static public List<Racer> CalculatePoints(List<Racer> toSort)
    {
        bool sorted = false;
        int endInt = 0;

        List<Racer> array = new List<Racer>(toSort);

        while (!sorted)
        {
            sorted = true;

            for (int i = 1; i < array.Count - endInt; i++)
            {
                if (array[i - 1].points < array[i].points)
                {
                    Racer holder = array[i - 1];
                    array[i - 1] = array[i];
                    array[i] = holder;
                    sorted = false;
                }
            }

            endInt++;
        }

        for(int i = 0; i < array.Count; i++)
        {
            array[i].overallPosition = i;
        }

        return array;

    }


    static private void Swap(List<Racer> array, int a, int b)
    {
        Racer holder = array[a];

        array[a] = array[b];
        array[b] = holder;

        return;
    }
}
